eth_utils/abi.py

```diff
@@ -3,10 +3,44 @@
 from .crypto import keccak
 
 
+def collapse_tuple_type(abi):
+    """Converts a tuple from a dict to a parenthesized list of its types.
+
+    >>> collapse_tuple_type(
+    ...     {
+    ...         'components': [
+    ...             {'name': 'anAddress', 'type': 'address'},
+    ...             {'name': 'anInt', 'type': 'uint256'},
+    ...             {'name': 'someBytes', 'type': 'bytes'},
+    ...         ],
+    ...         'type': 'tuple',
+    ...     }
+    ... )
+    '(address,uint256,bytes)'
+    """
+    if abi["type"] != "tuple":
+        return abi["type"]
+
+    component_types = [
+        collapse_tuple_type(component) for component in abi["components"]
+    ]
+
+    return "(" + ",".join(component_types) + ")"
+
+
+def get_fn_input_types(abi):
+    fn_input_types = [arg["type"] for arg in abi.get("inputs", [])]
+
+    for i in range(0, len(fn_input_types)):
+        if fn_input_types[i] == "tuple":
+            fn_input_types[i] = collapse_tuple_type(abi["inputs"][i])
+
+    return ",".join(fn_input_types)
```

Comment:
Here's a version that removes the list mutation

```python
from .functional import apply_to_return_value

@apply_to_return_value(",".join)
def _get_fn_input_types(abi):
    fn_arg_types = ...
   
    for arg_type in fn_arg_types:
            yield collapse_tuple_type(arg_type)
```

I believe this will be more performant.  I took the liberty to remove the inner `if/else` since `collapse_tuple_type` seems to appropriately handle non-tuple types.

Reply:
Thank you for pushing on this.  It definitely felt awkward before.  I made your suggestion, but then I took it one step further, and I'm quite happy with how simple it is now.  Let me know what you think.